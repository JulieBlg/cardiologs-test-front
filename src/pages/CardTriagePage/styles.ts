import { Theme } from '@material-ui/core';

const styles = (theme: Theme) => ({
    filtersBar: {
        marginTop: theme.spacing(3),
        marginBottom: theme.spacing(3),
    }
});

export default styles;