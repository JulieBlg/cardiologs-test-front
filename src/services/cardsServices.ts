import { Card } from '../typings';

interface CardStatus {
    card: Card;
    status: string;
}

export const getCards = async () => {
    const response = await window.fetch('http://localhost:4000/cards', {
        method: 'get',
        headers: { 'Content-Type': 'application/json' },
    });

    const data = await response.json();

    return data;
}

export const updateCardStatus = async ({card, status}: CardStatus) => {
    const response = await window.fetch(`http://localhost:4000/cards/${card.id}`, {
        method: 'put',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ ...card, status }),
    });

    const data = await response.json();

    return data;
}

export const searchCards = async (input: string) => {
    const response = await window.fetch(`http://localhost:4000/cards?q=${input}`, {
        method: 'get',
        headers: { 'Content-Type': 'application/json' },
    });

    const data = await response.json();

    return data;
}