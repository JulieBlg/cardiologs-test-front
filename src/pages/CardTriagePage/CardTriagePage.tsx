import React, { FC, useEffect, useState, ChangeEvent } from 'react';
import { Container, Grid, InputAdornment, TextField, makeStyles } from '@material-ui/core';
import { Search } from '@material-ui/icons';

import CardContainer from '../../containers/CardContainer';
import styles from './styles';
import { Card } from '../../typings';
import { getCards, updateCardStatus, searchCards } from '../../services/cardsServices';
import { statuses } from '../../global/statuses.json'

const CardTriagePage: FC = () => {
  const classes = makeStyles(styles)();
  const [cards, setCards] = useState([]);
  
  const getCardsByStatus = (status: string, cards: Card[]) => cards.filter(card => card.status === status);

  const handleClick = async (card: Card, status: string) => {
    await updateCardStatus({ card, status });

    await fetchData();
  }

  const fetchData = async () => {
    const data = await getCards();

    setCards(data);
  }

  const filterCards = async (event:ChangeEvent<HTMLInputElement>) => {
    const data = await searchCards(event.target.value);

    setCards(data);
  }

  useEffect(() => {
    fetchData();
  }, []);

  return (
    <Container maxWidth={false}>
      <Grid container spacing={3} justify="flex-end" className={classes.filtersBar}>
        <Grid item xs={3}>
          <TextField 
            id="filter" 
            label="Filter" 
            variant="outlined" 
            fullWidth
            onChange={filterCards}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">
                  <Search />
                </InputAdornment>
              ),
            }}
          />
        </Grid>
      </Grid>
      <Grid container spacing={3}>
        {statuses.map((status: string) => {
          const statusCards = getCardsByStatus(status, cards);
          return <CardContainer key={status} status={status} cards={statusCards} handleClick={handleClick} />
        })}
      </Grid>
    </Container>
  ); 
}

export default CardTriagePage;
