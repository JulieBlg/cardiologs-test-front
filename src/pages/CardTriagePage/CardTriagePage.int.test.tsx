import React from 'react';
import { render, getAllByTestId, getByText, fireEvent } from '@testing-library/react';

import CardTriagePage from './CardTriagePage';
import * as services from '../../services/cardsServices';

describe('CardTriagePage', () => {
  let getCardsSpy: jest.SpyInstance;
  let updateCardsSpy: jest.SpyInstance;
  let searchCardSpy: jest.SpyInstance;

  beforeAll(() => {
    getCardsSpy = jest
      .spyOn(services, 'getCards')
      .mockResolvedValue([
        {
            "arrhythmias": [
              "AFib",
              "AV Block",
              "Pause",
              "PSVC",
              "PVC"
            ],
            "created_date": "2020-03-10T13:14:59+0000",
            "id": 0,
            "patient_name": "Bob",
            "status": "PENDING"
          },
          {
            "arrhythmias": [
              "Pause"
            ],
            "created_date": "2020-01-01T00:12:21+0000",
            "id": 1,
            "patient_name": "Bill",
            "status": "COMPLETED"
          },
          {
            "arrhythmias": [
              "AFib",
              "Pause"
            ],
            "created_date": "2019-12-31T00:11:14+0000",
            "id": 2,
            "patient_name": "Elsa",
            "status": "DONE"
          },
      ]);

      updateCardsSpy = jest.spyOn(services, 'updateCardStatus');

      searchCardSpy = jest.spyOn(services, 'searchCards');
  });

  it('should render CardTriagePage properly', async () => {
    const { container, findByText } = render(<CardTriagePage />);

    expect(getCardsSpy).toHaveBeenCalled();

    await findByText('Patient ID : 0');

    expect(container).toMatchSnapshot();
  });

  it('should move card to another column', async () => {
    const { container, findByText, findAllByTestId } = render(<CardTriagePage />);

    expect(getCardsSpy).toHaveBeenCalled();

    await findByText('Patient ID : 0');

    const firstColumn = getAllByTestId(container, 'cardContainer')[0];
    var firstColumnCards = getAllByTestId(firstColumn, 'patientCard');
    expect(firstColumnCards).toHaveLength(1);

    const secondColumn = getAllByTestId(container, 'cardContainer')[1];
    var secondColumnCards = getAllByTestId(secondColumn, 'patientCard');
    expect(secondColumnCards).toHaveLength(1);

    const movingCardButton = getByText(firstColumnCards[0], 'Move Card To');
    movingCardButton.click();

    const options = await findAllByTestId('menuOption');
    options[0].click();

    expect(updateCardsSpy).toHaveBeenCalledWith(
      {
         "card":  {
           "arrhythmias":  [
             "AFib",
             "AV Block",
             "Pause",
             "PSVC",
             "PVC",
           ],
           "created_date": "2020-03-10T13:14:59+0000",
           "id": 0,
           "patient_name": "Bob",
           "status": "PENDING",
         },
         "status": "COMPLETED",
    });
  });

  it('should filter cards', () => {
    const { getByLabelText } = render(<CardTriagePage />);

    const input = getByLabelText('Filter');
    fireEvent.change(input, { target: { value: 'Pause'}});

    expect(searchCardSpy).toHaveBeenCalledWith('Pause');
  });
})

