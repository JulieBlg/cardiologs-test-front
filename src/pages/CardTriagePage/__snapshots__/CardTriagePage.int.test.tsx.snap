// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTriagePage should render CardTriagePage properly 1`] = `
<div>
  <div
    class="MuiContainer-root"
  >
    <div
      class="MuiGrid-root makeStyles-filtersBar-255 MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-justify-xs-flex-end"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="filter"
            id="filter-label"
          >
            Filter
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              id="filter"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-187 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-189"
              >
                <span>
                  Filter
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
      >
        <div
          class="MuiPaper-root makeStyles-patientCardContainer-256 MuiPaper-elevation2 MuiPaper-rounded"
          data-testid="cardContainer"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          >
            PENDING
          </h2>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
            data-testid="patientCard"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiPaper-root MuiCard-root makeStyles-patientCard-258 MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiCardHeader-root"
                >
                  <div
                    class="MuiCardHeader-avatar"
                  >
                    <div
                      class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
                    >
                      B
                    </div>
                  </div>
                  <div
                    class="MuiCardHeader-content"
                  >
                    <span
                      class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
                    >
                      Bob
                    </span>
                    <span
                      class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      Patient ID : 0
                    </span>
                  </div>
                  <div
                    class="MuiCardHeader-action"
                  >
                    <div>
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-menuButton-277 MuiButton-containedSizeSmall MuiButton-sizeSmall"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          Move Card To
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root"
                            focusable="false"
                            role="presentation"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M10 17l5-5-5-5v10z"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiCardContent-root"
                >
                  <p
                    class="MuiTypography-root makeStyles-listTitle-259 MuiTypography-body1"
                  >
                    Arrhythmias :
                  </p>
                  <div
                    class="MuiGrid-root makeStyles-arrythmiasList-260 MuiGrid-item MuiGrid-grid-xs-12"
                  >
                    <div
                      class="MuiChip-root makeStyles-arrythmia-261"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        AFib
                      </span>
                    </div>
                    <div
                      class="MuiChip-root makeStyles-arrythmia-261"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        AV Block
                      </span>
                    </div>
                    <div
                      class="MuiChip-root makeStyles-arrythmia-261"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        Pause
                      </span>
                    </div>
                    <div
                      class="MuiChip-root makeStyles-arrythmia-261"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        PSVC
                      </span>
                    </div>
                    <div
                      class="MuiChip-root makeStyles-arrythmia-261"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        PVC
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
      >
        <div
          class="MuiPaper-root makeStyles-patientCardContainer-345 MuiPaper-elevation2 MuiPaper-rounded"
          data-testid="cardContainer"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          >
            COMPLETED
          </h2>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
            data-testid="patientCard"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiPaper-root MuiCard-root makeStyles-patientCard-347 MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiCardHeader-root"
                >
                  <div
                    class="MuiCardHeader-avatar"
                  >
                    <div
                      class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
                    >
                      B
                    </div>
                  </div>
                  <div
                    class="MuiCardHeader-content"
                  >
                    <span
                      class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
                    >
                      Bill
                    </span>
                    <span
                      class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      Patient ID : 1
                    </span>
                  </div>
                  <div
                    class="MuiCardHeader-action"
                  >
                    <div>
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-menuButton-352 MuiButton-containedSizeSmall MuiButton-sizeSmall"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          Move Card To
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root"
                            focusable="false"
                            role="presentation"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M10 17l5-5-5-5v10z"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiCardContent-root"
                >
                  <p
                    class="MuiTypography-root makeStyles-listTitle-348 MuiTypography-body1"
                  >
                    Arrhythmias :
                  </p>
                  <div
                    class="MuiGrid-root makeStyles-arrythmiasList-349 MuiGrid-item MuiGrid-grid-xs-12"
                  >
                    <div
                      class="MuiChip-root makeStyles-arrythmia-350"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        Pause
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
      >
        <div
          class="MuiPaper-root makeStyles-patientCardContainer-353 MuiPaper-elevation2 MuiPaper-rounded"
          data-testid="cardContainer"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          >
            DONE
          </h2>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
            data-testid="patientCard"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiPaper-root MuiCard-root makeStyles-patientCard-355 MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiCardHeader-root"
                >
                  <div
                    class="MuiCardHeader-avatar"
                  >
                    <div
                      class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
                    >
                      E
                    </div>
                  </div>
                  <div
                    class="MuiCardHeader-content"
                  >
                    <span
                      class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
                    >
                      Elsa
                    </span>
                    <span
                      class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      Patient ID : 2
                    </span>
                  </div>
                  <div
                    class="MuiCardHeader-action"
                  >
                    <div>
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-menuButton-360 MuiButton-containedSizeSmall MuiButton-sizeSmall"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          Move Card To
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root"
                            focusable="false"
                            role="presentation"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M10 17l5-5-5-5v10z"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiCardContent-root"
                >
                  <p
                    class="MuiTypography-root makeStyles-listTitle-356 MuiTypography-body1"
                  >
                    Arrhythmias :
                  </p>
                  <div
                    class="MuiGrid-root makeStyles-arrythmiasList-357 MuiGrid-item MuiGrid-grid-xs-12"
                  >
                    <div
                      class="MuiChip-root makeStyles-arrythmia-358"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        AFib
                      </span>
                    </div>
                    <div
                      class="MuiChip-root makeStyles-arrythmia-358"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        Pause
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
